fast_matched_filter/src/matched_filter.c

```diff
@@ -112,29 +120,55 @@ float corrc(float *templates, float sum_square_template,
     for (i = 0; i < n_samples_template; i++){
         numerator += templates[i] * data[i];
     }
-    denominator = sum_square_template * csum_square_data[0];
+    denominator = sum_square_template * (float)(csum_square_data[n_samples_template] - csum_square_data[-1]);
 
     if (denominator > STABILITY_THRESHOLD) cc = numerator / sqrt(denominator);
 
     return cc;
 }
 
 //-------------------------------------------------------------------------
-void csum(double *data_sq, int n_samples_template, int n_samples_data,
-          int n_stations, int n_components,
-          double *csum_square_data) {
-
-    int ch, i, n;
-    for (ch = 0; ch < (n_stations*n_components); ch++){
-        double *csum_ch = NULL, *data_sq_ch = NULL;
-        csum_ch = csum_square_data + ch * n_samples_data;
-        data_sq_ch = data_sq + ch * n_samples_data;
-
-        // sliding cumulative sum
-#pragma omp parallel for private(ch, i, n) shared(data_sq_ch, csum_ch)
-        for (n = 0; n < n_samples_data-n_samples_template; n++){
-            for (i = 0; i < n_samples_template; i++) csum_ch[n] += data_sq_ch[n+i];
+void cumsum_square_data(float *data, int n_samples_data, float *weights,
+                        int n_stations, int n_components,
+                        double *csum_square_data) {
+    int s, c;
+    int station_offset, component_offset, d;
+
+    for (s = 0; s < n_stations; s++) {
+        station_offset = s * n_components;
+
+        for (c = 0; c < n_components; c++) {
+            component_offset = station_offset + c;
+            if (weights[component_offset] == 0) continue;
```

Comment:
It's not correct to check this condition, because weights is not the same for different templates.

Reply:
Good point.